### PR TITLE
Access Transformer for Inspectors

### PR DESCRIPTION
**Pull request by markuspf**
### Scope & Purpose

My amateur attempt at making (de)serialising Enums into velocypack with Inspectors possible.

As a result I concocted a way to specifiy a transformation from a type into a storage type by way of an `Access`-transformer: Suppose you have a type `T` but really would like to store it as a type `StorageT`, which also happens to have an inspector defined already. 

With this PR you can create a storage class

```
template<typename Enum>
struct EnumStorage {
  using MemoryType = Enum;

  std::underlying_type_t<Enum> code;
  std::string message;

  explicit EnumStorage(Enum e)
      : code(static_cast<std::underlying_type_t<Enum>>(e)),
        message(to_string(e)){};
  explicit EnumStorage() {}

  operator Enum() const { return Enum(code); }
};

template<class Inspector, class Enum>
auto inspect(Inspector& f, EnumStorage<Enum>& e) {
  return f.object(e).fields(f.field("code", e.code),
                            f.field("message", e.message).fallback(""));
}
```
and instruct inspectors to use this transformers like so

```
template<>
struct Access<AnEnumClass>
    : StorageTransformerAccess<AnEnumClass, EnumStorage<AnEnumClass>> {};
```

 
- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [x] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*
  - [ ] Backport for 3.7: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

**Review by maierlars — APPROVED** [review]
Like it

File: tests/Basics/InspectionTest.cpp [hunk 1]
```diff
@@ -1472,4 +1514,33 @@ TEST_F(VPackInspectionTest, deserialize_throws) {
...
+    velocypack::Builder builder;
+
+    AnEnumClass const d = AnEnumClass::Option1;
+    arangodb::velocypack::serialize(builder, d);
+
+    velocypack::Slice slice = builder.slice();
+    ASSERT_TRUE(slice.isObject());
+    EXPECT_EQ(static_cast<std::underlying_type_t<AnEnumClass>>(d),
+              slice["code"].getInt());
+  }
```
Comment (mpoeter) [inline tests/Basics/InspectionTest.cpp:1528]: Shouldn't we also add an assertion for the message?

File: tests/Basics/InspectionTest.cpp [hunk 2]
```diff
@@ -298,6 +300,42 @@ struct Specialization {
...
+        message(to_string(e)){};
+  explicit EnumStorage() {}
+
+  operator Enum() const { return Enum(code); }
+};
+
+template<class Inspector, class Enum>
+auto inspect(Inspector& f, EnumStorage<Enum>& e) {
+  return f.object(e).fields(f.field("code", e.code),
+                            f.field("message", e.message).fallback(""));
```
Comment (mpoeter) [inline tests/Basics/InspectionTest.cpp:336]: This works but actually loads the message just so we discard it afterwards. Alternatively you could do it like this:
```
if constexpr (Inspector::isLoading) {
  return f.object(e).fields(f.field("code", e.code),
                            f.ignoreField("message");
} else {
  return f.object(e).fields(f.field("code", e.code),
                            f.field("message", e.message));
}
```

File: tests/Basics/InspectionTest.cpp [hunk 3]
```diff
@@ -1472,4 +1514,33 @@ TEST_F(VPackInspectionTest, deserialize_throws) {
...
+
+  {
+    auto const expected = AnEnumClass::Option3;
+    velocypack::Builder builder;
+
+    builder.openObject();
+    builder.add(
+        "code",
+        VPackValue(static_cast<std::underlying_type_t<AnEnumClass>>(expected)));
+    builder.close();
```
Comment (mpoeter) [inline tests/Basics/InspectionTest.cpp:1538]: Here we should also add a message to the object, so we sure it is ignored correctly.

**Review by mpoeter — APPROVED** [review]

**Pull request merged**